### ISSUE TITLE
Problem extending `only`directive

### ISSUE THREAD

**Issue by gabrielfalcao** [sphinx-doc/sphinx#5555]
TL;DR: this is a follow-up to #4374 we read the documentation of sphinx extensions, docutils and even explored the sphinx source code itself.

The directive ``.. only::`` cannot be generated from another directive.

Sphinx versions used 1.8.1 and master (2c244f9b48315d19eb61ea006a99ffad54c2f83c)

## The problem to solve

```rst
.. only:: internal-only

   .. rst-class:: myclass

   .. container::

      Our documentation is full of "internal-only" sections but wrapping such parts with
      ``.. container::`` is overly verbose and error-prone.

      .. important:: 

         we have several *internal-only* toctrees and other sections,
         such as below.

      .. toctree::
         :caption: Toctree test
         :maxdepth: 1

         docs/test.rst

      Title test
      ----------

      Blabla.
```

Using the `.. rst-class::` directive will cause the HTML renderer to add the CSS class to all its children.
We want to wrap all children of `.. only::` into a single parent, so the CSS class is added only to it, so we use the ``.. container::`` directive.

## The final use case

We followed @tk0miya's [suggestion](https://github.com/sphinx-doc/sphinx/issues/4374#issuecomment-405652865) and wrote our own extension. It's basically
a directive called `conditional_output` which builds a subtree of nodes to reproduce what the example above shows but without the need to use 3 directives (``only``, ``rst-class`` and ``container``).

The directive has 2 parameters:

- ``class``: forwards the parameter to ``rst-class``.
- ``wrapper``: when present, we wrap the content of the directive with a container div. It replaces ``container``.

**Example of use:**

```rst
.. conditional_output:: internal-only
   :class: myclass
   :wrapper:

   Our documentation is full of "internal-only" sections but wrapping such parts with ``.. container:: is overly verbose and error-prone.

   .. important:: we have several *internal-only* toctrees, such as below.

   .. toctree::
      :caption: Toctree test
      :maxdepth: 1

      docs/test.rst

   Title test
   ----------

   Blabla.
```

## The problem

The first version of the extension causes docutils to inject a `system_message` in our doctree with the following warning:

```bash
Title test
----------
WARNING: Unexpected section title.
```

To fix this warning we created [this patch](https://github.com/NewStore/sphinx/pull/3/files#diff-a6c71e8f1522f9e3cacceec1fe7df73cR324).

This patch takes care of the warning but now we have a [traceback](https://gist.github.com/gabrielfalcao/61e39bbfe2d2eb67443da3ab9293a0ad#file-sphinx-traceback).

For more detail, here is the [extension code](https://gist.github.com/gabrielfalcao/61e39bbfe2d2eb67443da3ab9293a0ad#file-sphinx_ext_conditional_output-py).

Considering that we already opened [this ticket](#4374), and went through the hardship of learning Sphinx and Docutils
could we get some specific pointers or code samples on how to create such an extension?

We will also contribute to the docs by created more tutorials about extensions. It's really not that easy!

----

### Replicate all logic of `sphinx.directives.other.Only` in our own directive + transformer

The initial extension looked more-or-less like [this](https://gist.github.com/gabrielfalcao/1bcb24a1db811b6025415455b329b39a)

But toctrees were not being processed, we realized that multiple parts of the sphinx codebase hard-codes the use of `addnodes.only` for both traversal and node creation:

- [sphinx/directives/other.py](https://github.com/sphinx-doc/sphinx/blob/2c244f9b48315d19eb61ea006a99ffad54c2f83c/sphinx/directives/other.py#L328)
- sphinx/util/nodes.py:
  - [TocTreeCollector.process_doc](https://github.com/sphinx-doc/sphinx/blob/2c244f9b48315d19eb61ea006a99ffad54c2f83c/sphinx/environment/collectors/toctree.py#L87)
  - [TocTreeCollector.assign_section_numbers](https://github.com/sphinx-doc/sphinx/blob/2c244f9b48315d19eb61ea006a99ffad54c2f83c/sphinx/environment/collectors/toctree.py#L162)

So, to decouple the use of `addnodes.only` in sphinx we tried 2 different approaches:

- [extend by inheritance](https://github.com/NewStore/sphinx/pull/2/files#diff-e32fc687e1850d2e91f8fc402f480cbfR323)
- [introduce a new event to support upgrading nodes](https://github.com/NewStore/sphinx/pull/1)

But both approaches seemed too contrived and did not solve our problem.

**Comment by gabrielfalcao** [comment]
@tk0miya is this a bug on sphinx or something we did wrong?

**Comment by tk0miya** [comment]
Sorry for late.

It seems the error was raised on `FilterSystemMessages`.
https://github.com/sphinx-doc/sphinx/blob/9c2fd94204fce9f056104083f20f094827619781/sphinx/transforms/__init__.py#L311-L321

It removes `system_message` node from doctree. But the error says your doctree is broken; a parent of `system_message` node does not know the node. I couldn't find where such broken tree is generated.

**Comment by tk0miya** [comment]
I'm closing this because of not a bug of Sphinx.
Thanks,
